ffi/src/lib.rs

```diff
@@ -8,6 +8,66 @@
 // CONDITIONS OF ANY KIND, either express or implied. See the License for the
 // specific language governing permissions and limitations under the License.
 
+//! This module exposes an Foreign Function Interface (FFI) that allows Mentat to be
+//! called from other languages.
+//!
+//! Functions that are available to other languages in this module are defined as
+//! extern "C" functions which allow them to be layed out correctly for the
+//! platform's C ABI. They all have a `#[no_mangle]` decorator to ensure
+//! Rust's name mangling is turned off, so that it is easier to link to.
+//!
+//! Mentat's FFI contains unsafe code. As it is an interface between foreign code
+//! and native Rust code, Rust cannot guarentee that the types and data that it is
+//! handling that have been passed to it from another language are present and in
+//! the format it is expecting.
+//! This interface is designed to ensure that nothing unsafe ever leaves this module
+//! and enters Mentat proper.
+//!
+//! Structs defined with `#[repr(C)]` are guaranteed to have a layout that is compatible
+//! with the platform's representation in C.
+//!
+//! This API passes pointers in two ways, depending on the lifetime of the object and
+//! what object owns it.
+//! Pointers to values that can guarenteed to live beyond the lifetime of the function,
+//! are passed over the FFI as a raw pointer.
+//! ```
+//! value as *const TypedValue
+//! ```
+//! Pointers to values that cannot be guarenteed to live beyond the lifetime of the function
+//! are first Boxed so that they live on the heap, and the raw pointer passed this way.
+//! ```
+//! Box::into_raw(Box::new(value))
+//! ```
+//!
+//! The memory for values that are moved onto the heap before being passed over the FFI
+//! are no longer managed by Rust, but Rust still owns the value. Therefore the pointer
+//! must be returned to Rust in order to be released. To this effect a number of `destructor`
+//! function are provided for each Rust value type that is passed, and a catch all destructor
+//! to destroy memory for `#[repr(C)]` values. The destructors reclaim the memory via `Box`
+//! and then drop the reference, causing the memory to be released.
+//!
+//! A macro has been provided to make defining destructors easier.
+//! ```
+//! define_destructor!(query_builder_destroy, QueryBuilder);
+//! ```
+//!
+//! Passing a pointer to released memory will cause Mentat to crash, so callers have to be
+//! careful to ensure they manage their pointers properly. Failure to call a destructor for a
+//! value on the heap will cause a memory leak.
+//!
+//! Most of the functions exposed in this FFI have a direct mapping to existing Mentat APIs.
+//! Application logic has been kept to a minumum in order to provide the greatest flexibility
+//! for callers using the interface, however there are a one exception where several steps
```

Comment:
nit: "there are a one"...

Rather than fixing the grammar, I think we should say something more general: "This module is generally a direct mapping to existing Mentat APIs, because [rationale].  However, in some instances a single convenience ... because ... [rationale for the class of exceptional cases]."

Let's not have this be out of date quite so quickly!